test/ha/kourier_gateway_test.go

```diff
@@ -0,0 +1,139 @@
+// +build e2e
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"log"
+	"sort"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/ptr"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/conformance/ingress"
+	"knative.dev/serving/test/e2e"
+)
+
+const (
+	kourierGatewayDeployment = "3scale-kourier-gateway"
+	kourierGatewayLabel      = "app=3scale-kourier-gateway"
+	kourierService           = "kourier"
+	minProbes                = 100  // We want to send at least 100 requests.
+	SLO                      = 0.99 // We permit 0.01 of requests to fail due to killing the Gateway.
+)
+
+// The Kourier Gateway does not have leader election enabled.
+// The test ensures that stopping one of the gateway pods doesn't affect user applications.
+func TestKourierGatewayHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	if err := pkgTest.WaitForDeploymentScale(clients.KubeClient, kourierGatewayDeployment, ingressNamespace, haReplicas); err != nil {
+		t.Fatalf("Deployment %s not scaled to %d: %v", kourierGatewayDeployment, haReplicas, err)
+	}
+
+	svcName, svcPort, svcCancel := ingress.CreateRuntimeService(t, clients, networking.ServicePortNameHTTP1)
+	defer svcCancel()
+
+	// Create a service that we will continually probe while restarting Kourier gateway.
+	_, _, ingressCancel := ingress.CreateIngressReady(t, clients, createIngressSpec(svcName, svcPort))
+	defer ingressCancel()
+
+	url := apis.HTTP(svcName + domain)
+	prober := test.NewProberManager(log.Printf, clients, minProbes)
+	// Start the prober and let it send requests while restarting both gateway pods.
+	prober.Spawn(url.URL())
+	defer assertSLO(t, prober)
+
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(ingressNamespace).List(metav1.ListOptions{
+		LabelSelector: kourierGatewayLabel,
+	})
+	if err != nil {
+		t.Fatal("Failed to get Gateway pods:", err)
+	}
+	gatewayPod := pods.Items[0].Name
+
+	kourierServiceOpts := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("metadata.name", kourierService).String(),
```

Comment:
This looks a little weird 🤔. I think we should just use a `Get` to get exactly the service. Why do we need to list them?

Reply:
The goal is to have just one reusable function in knative.dev/pkg. But we can have two separate (more specific) functions.

Reply:
In knative/serving we use a little different selector. Not just the name of the service. So I wanted to create a common function that could be used for kourier and serving as well.